cmd/kubeless/function.go

```diff
@@ -100,14 +119,246 @@ func parseMemory(mem string) (resource.Quantity, error) {
+			logrus.Fatal(err)
+		}
+		pfo := k8scmd.PortForwardOptions{
+			RESTClient: k8sClient,
+			Namespace:  "kubeless",
+			Config:     k8sClientConfig,
+			PodName:    podName,
+			PodClient:  clientset.Core(),
+			Ports:      portSlice,
+			PortForwarder: &defaultPortForwarder{
+				cmdOut: devNull,
+				cmdErr: os.Stderr,
+			},
+			StopChannel:  make(chan struct{}, 1),
+			ReadyChannel: make(chan struct{}),
+		}
+		err = pfo.RunPortForward()
+		if err != nil {
+			logrus.Fatalf("Connection failed: %v", err)
+		}
+
+	}()
+	return
+}
+
+func getMinioURLFromOutsideCluster() (url string, err error) {
+	f := cmdutil.NewFactory(nil)
+	k8sClient, err := f.RESTClient()
+	if err != nil {
+		return
+	}
+
+	pods := &v1.PodList{}
+	pod := v1.Pod{}
+	k8sClient.Get().Namespace("kubeless").Resource("pods").Do().Into(pods)
+	for i := range pods.Items {
+		if pods.Items[i].Labels["kubeless"] == "minio" {
+			pod = pods.Items[i]
+			break
+		}
+	}
+
+	port, err := getLocalPort()
+	if err != nil {
+		return
+	}
+
+	// Minio requires to serve content at "root" eg. minio.com:9000
+	// so instead of using the Kubernetes Proxy we need to run
+	// a port forward to Minio's pod
+	runPortForward(f, k8sClient, pod.Name, port)
+	err = waitForHTTPServer("localhost", port)
+	if err != nil {
+		return
+	}
+	url = "localhost:" + port
+	return
+}
+
+func getMinioObject(minioClient minio.Client, bucket, name string) (object []byte, err error) {
+	var dir string
+	dir, err = ioutil.TempDir("", "")
+	if err != nil {
+		return
+	}
+	defer os.RemoveAll(dir)
+
+	err = minioClient.FGetObject(bucket, name, path.Join(dir, name), minio.GetObjectOptions{})
+	if err != nil {
+		return
+	}
+	object, err = ioutil.ReadFile(path.Join(dir, name))
+	return
+}
+
+func getMinioClient(endpoint string) (minioClient *minio.Client, err error) {
+	cli, err := utils.GetRestClientOutOfCluster("", "v1", "/api")
+	if err != nil {
+		return
+	}
+	result := &v1.Secret{}
+	cli.Get().
+		Namespace("kubeless").
+		Resource("secrets").
+		Name("minio-key").
+		Do().
+		Into(result)
+
+	accessKeyID := string(result.Data["accesskey"])
+	secretAccessKey := string(result.Data["secretkey"])
+	useSSL := false
+	minioClient, err = minio.New(endpoint, accessKeyID, secretAccessKey, useSSL)
+	return
+}
+
+func uploadFunction(file string) (checksum string, err error) {
+
+	stats, err := os.Stat(file)
+	if stats.Size() > int64(52428800) { // TODO: Make the max file size configurable
+		err = errors.New("The maximum size of a function is 50MB")
+		return
+	}
+	sha256Sum, checksum, err := getFileSha256(file)
+
+	endpoint, err := getMinioURLFromOutsideCluster()
+	if err != nil {
+		return
+	}
+	minioClient, err := getMinioClient(endpoint)
+	if err != nil {
+		return
+	}
+	bucketName := "functions"
+	objectName := path.Base(file) + "." + checksum
+
+	_, getObjectErr := minioClient.StatObject(bucketName, objectName, minio.StatObjectOptions{})
+	if getObjectErr == nil {
+		// File already exists, validate checksum
```

Comment:
The checksum is in the filename, so what is the purpose of downloading it to check that it has that same checksum?